src/operator/numpy/np_where_op-inl.h

```diff
@@ -0,0 +1,218 @@
+    const int ndim_delta = expanded_ishape.ndim() - ishape.ndim();
+    for (int j = 0; j < ishape.ndim(); ++j) {
+      expanded_ishape[j + ndim_delta] = ishape[j];
+    }
+    in_strides[i] = mxnet_op::calc_stride(expanded_ishape.get<broadcast::MAX_DIM>());
+  }
+  TShape expanded_oshape(broadcast::MAX_DIM, 1);
+  const int ndim_delta = expanded_oshape.ndim() - outputs[0].shape_.ndim();
+  for (int j = 0; j < outputs[0].shape_.ndim(); ++j) {
+    expanded_oshape[j + ndim_delta] = (outputs[0].shape_)[j];
+  }
+  Shape<broadcast::MAX_DIM> oshape = expanded_oshape.get<broadcast::MAX_DIM>();
+  MSHADOW_TYPE_SWITCH_WITH_BOOL(outputs[0].type_flag_, DType, {
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(inputs[0].type_flag_, CType, {
+      mxnet_op::Kernel<numpy_where_kernel<broadcast::MAX_DIM>, xpu>::Launch(
+        s, outputs[0].Size(), req[0],
+        in_strides[0], in_strides[1], in_strides[2], oshape,
+        inputs[0].dptr<CType>(), inputs[1].dptr<DType>(),
+        inputs[2].dptr<DType>(), outputs[0].dptr<DType>());
+    });
+  });
+}
+
+template<typename xpu>
+inline void NumpyWhereOpBackward(const nnvm::NodeAttrs& attrs,
+                                 const OpContext& ctx,
+                                 const std::vector<TBlob>& inputs,
+                                 const std::vector<OpReqType>& req,
+                                 const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 2U);
+  CHECK(common::is_float(inputs[0].type_flag_)) << "Backward only supports float types!";
+  if (inputs[0].shape_.Size() == 0U) return;  // zero-size tensor
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  // get expanded oshape
+  TShape expanded_oshape(broadcast::MAX_DIM, 1);
+  int ndim_delta = expanded_oshape.ndim() - inputs[0].shape_.ndim();
+  for (int j = 0; j < inputs[0].shape_.ndim(); ++j) {
+    expanded_oshape[j + ndim_delta] = (inputs[0].shape_)[j];
+  }
+  Shape<broadcast::MAX_DIM> oshape = expanded_oshape.get<broadcast::MAX_DIM>();
+  // get cond stride
+  TShape expanded_cshape(broadcast::MAX_DIM, 1);
+  ndim_delta = expanded_cshape.ndim() - inputs[1].shape_.ndim();
+  for (int j = 0; j < inputs[1].shape_.ndim(); ++j) {
+    expanded_cshape[j + ndim_delta] = (inputs[1].shape_)[j];
+  }
+  Shape<broadcast::MAX_DIM> cstride =
+      mxnet_op::calc_stride(expanded_cshape.get<broadcast::MAX_DIM>());
+  // get expanded lshape
+  TShape expanded_lshape(broadcast::MAX_DIM, 1);
+  ndim_delta = expanded_lshape.ndim() - outputs[0].shape_.ndim();
+  for (int j = 0; j < outputs[0].shape_.ndim(); ++j) {
+    expanded_lshape[j + ndim_delta] = (outputs[0].shape_)[j];
+  }
+  // get expanded rshape
+  TShape expanded_rshape(broadcast::MAX_DIM, 1);
+  ndim_delta = expanded_rshape.ndim() - outputs[1].shape_.ndim();
+  for (int j = 0; j < outputs[1].shape_.ndim(); ++j) {
+    expanded_rshape[j + ndim_delta] = (outputs[1].shape_)[j];
+  }
+
+  MSHADOW_TYPE_SWITCH_WITH_BOOL(inputs[0].type_flag_, DType, {
```

Comment:
Consider do:
```
const TBlob& cond = inputs[0];
const TBlob& x = inputs[1];
const TBlob& y = inputs[2];
const TBlob& out = outputs[0];
```
and substitute all the `inputs[i]` `outputs[j]` to improve readability.